Add Telemetry to `keystone dev` (#8118)
keystonejs/keystone: docs/pages/docs/reference/telemetry.md

Please note that Keystone currently uses [NextJS](https://nextjs.org/) and [Prisma](https://www.prisma.io/) as software dependencies, and each of these projects include their own telemetry with their own respective privacy policies.
You can learn more about them by visiting the following:

- [https://nextjs.org/telemetry](https://nextjs.org/telemetry)
- [https://www.prisma.io/docs/concepts/more/telemetry](https://www.prisma.io/docs/concepts/more/telemetry)

---

## Why do we collect and hold telemetry?

Keystone only collects telemetry that we reasonably require for the continued development of the software, and, to the extent applicable, we do so in accordance with the Australian Privacy Principles.

Telemetry helps us learn about how Keystone is being used by developers and projects that may not be actively participating on our GitHub, Twitter and Slack.
The questions we want to currently answer using telemetry are the following:

> - How many active developers are using Keystone, compared to how many are active in the community

Author: ```suggestion
- How many active developers are using Keystone, compared to how many are active in the community,
```

---

> title: "Privacy Policy (Telemetry)"

Reviewer: Why do we want to title it as **Privacy Policy** instead of just **Telemetry**?

Author: Because it is also our Privacy Policy

> description: "Keystone Collects anonymous telemetry events when running `keystone dev` - Learn what is sent, what the information is used for, and how to opt-out"

Reviewer: ```suggestion
description: "Keystone collects anonymous telemetry events when running `keystone dev` - Learn what is sent, what the information is used for, and how to opt-out"
```

---

> ## Summary

Reviewer: ```suggestion

```
We can start directly with content without a title unless you think this title is important

> ### Summary

Reviewer: ```suggestion
## Summary
```

This and all other top level headings in this page to use h2 instead of h3.

This page describes how we collect and hold **anonymous** telemetry data, the rationale for why we are collecting this information, when it is collected, and how **you can opt-out**. Unless you opt out, you consent to us collecting, holding, analysing and using the telemetry data in accordance with this policy, which may be updated from time to time.

Please note that Keystone currently uses [NextJS](https://nextjs.org/) and [Prisma](https://www.prisma.io/) as software dependencies, and each of these projects include their own telemetry with their own respective privacy policies.
You can learn more about them by visiting the following:

- [https://nextjs.org/telemetry](https://nextjs.org/telemetry)
- [https://www.prisma.io/docs/concepts/more/telemetry](https://www.prisma.io/docs/concepts/more/telemetry)

---

## Why do we collect and hold telemetry?

Keystone only collects telemetry that we reasonably require for the continued development of the software, and, to the extent applicable, we do so in accordance with the Australian Privacy Principles.

> Telemetry helps us learn about how Keystone is being used by developers and projects that may not be actively participating on our GitHub, Twitter and slack.

Reviewer: ```suggestion
Telemetry helps us learn about how Keystone is being used by developers and projects that may not be actively participating on our GitHub, Twitter and Slack.
```

The questions we want to currently answer using telemetry are the following:

> - How many active developers are using Keystone, compared to how many are active in the community

Reviewer: ```suggestion
- How many active developers are using Keystone vs how many are active in the community?
```

> - How complex are projects that use Keystone,

Reviewer: ```suggestion
- How complex are projects that use Keystone?
```

> - What field types and `@keystone-6` or community packages are being used by developers, and

Reviewer: ```suggestion
- What field types, `@keystone-6` packages or community packages are being used by developers?
```

> - What operating systems and node versions are being used by the community

Reviewer: ```suggestion
- What operating systems and node versions are being used by the community?
```

There are four scenarios when running `keystone dev`, where we **do not collect** telemetry:

- The first time running `keystone dev`, when displaying the telemetry notice, or
- If the `CI` environment variable is set (to anything), or
- If the `NODE_ENV` environment variable is set to `"production"`, or
- If you have opted out - see [How to opt-out below](#how-to-opt-out)

In any of these scenarios, telemetry will be disabled and prevented from sending any information.

---

## What information is collected?

{% hint kind="warn" %}
⚠️ The exact structure and contents of the telemetry information collected may be subject to change when the software is updated.
Developers should watch [our releases on GitHub](https://github.com/keystonejs/keystone/releases) for any changes in what telemetry information is reported by the Software.
Any material increase in the amount of information collected will result in us re-notifying developers of that change the next time they run `keystone dev`; and updating this page.

{% /hint %}

Keystone collects telemetry information in the form of two different types of data reports:

> - Information about the device running `keystone dev`, and

Reviewer: ```suggestion
- Information about the device running `keystone dev`
```

- Information about the project’s configuration

We refer to these two different reports, as “device telemetry” and “project telemetry” respectively.

These reports are forwarded to [https://telemetry.keystonejs.com/](https://telemetry.keystonejs.com/), and are reported separately to minimize any correlation between them insofar as the timing and grouping of that data, that an otherwise combined report may have. We are collecting these two reports for different reasons, and thus have no need to associate them.

We additionally record a timestamp of the time that the report is received by the server at [https://telemetry.keystonejs.com](https://telemetry.keystonejs.com/).

**Device Telemetry**
The type of information contained within a device telemetry report is currently:

> - The last date you used `keystone dev`, and

Reviewer: ```suggestion
- The last date you used `keystone dev`
```

> - The node major version number, and

Reviewer: ```suggestion
- The node major version number
```

- The name of your operating system

A device telemetry report is formatted as JSON and currently looks like:

```json
{
  "previous": "2022-11-23",
  "os": "darwin",
  "node": "18"
}
```

**Project Telemetry**

The type of information contained within a project telemetry report is currently:

> - The last date you used `keystone dev` for this project, and

Reviewer: ```suggestion
- The last date you used `keystone dev` for this project
```

> - The resolved versions of any `@keystone-6` packages used by this project, and

Reviewer: ```suggestion
- The resolved versions of any `@keystone-6` packages used by this project
```

> - The number of lists for this project, and

Reviewer: ```suggestion
- The number of lists for this project
```
